### PR TITLE
Add "prepare" and "check" watchers

### LINKED ISSUE #710: Loop duration and wakeup delay statistics
We (Envoy; https://github.com/envoyproxy/envoy) would like to start collecting histograms on the epoll (or more generally, event loop) duration. This will help us to identify badly behaving event handlers in extensions to Envoy. In addition, we'd like to track wakeup delay; specifically the time between when the event manager blocks for the next wakeup and when the wakeup occurs, to get an idea of kernel load.

The existing libevent event loop (http://www.wangafu.net/~nickm/libevent-book/Ref3_eventloop.html) doesn't really support this, even if we run with EVLOOP_ONCE.

I'm interested in hearing from libevent folks how we could best add this support; we don't mind taking on this work item but would like some guidance on the best fit to libevent's API. I think libev/libuv could support this with their check/prepare/idle watchers, but maybe there is something lighterweight to add to libevent.

CC @mattklein123
Related discussion in https://github.com/envoyproxy/envoy/issues/4952

### PR DESCRIPTION
**Pull request by mergeconflict**
Adds two new callbacks: "prepare" watchers, which fire immediately before we poll for I/O, and "check" watchers, which fire immediately after we finish polling and before we process events. This allows other event loops to be embedded into libevent's, and enables certain performance monitoring.

Closes: #710

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -737,6 +737,7 @@ set(HDR_PUBLIC
     include/event2/event.h
     include/event2/event_compat.h
     include/event2/event_struct.h
+    include/event2/event_watcher.h
```
Comment (mergeconflict) [inline CMakeLists.txt]: I think there are non-CMake builds I need to support as well... can somebody help direct me to the other build files I need to update? Thanks!

File: evwatcher.c [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+event_watcher_base(struct event_watcher *watcher)
+{
+	return watcher->base;
+}
+
+void
+event_watcher_free(struct event_watcher *watcher)
+{
+	TAILQ_REMOVE(&watcher->base->event_watchers[watcher->type], watcher, next);
+	free(watcher);
```
Comment (mergeconflict) [inline evwatcher.c]: Oops, I think this should be `mm_free`.

File: include/event2/event_watcher.h [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	this as a struct rather than individual parameters to the callback function for the sake of future
+	extensibility.
+ */
+struct prepare_watcher_cb_info {
+	/**
+		The current time (may be cached). This value is scoped to this "prepare" callback; if you need
+		to retain it outside the scope of the callback (e.g. to measure the duration between callbacks),
+		make a copy.
+	 */
+	struct timeval now;
```
Comment (mergeconflict) [inline include/event2/event_watcher.h]: Question: is it ok to include this here, or do we prefer users to use `event_base_update_cache_time` and `event_base_gettimeofday_cached`?

File: include/event2/event_watcher.h [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+	/**
+		The timeout duration passed to the underlying implementation's `dispatch`. This will be NULL if
+		there are no pending EV_TIMEOUT events. It may be a useful performance statistic to compare this
+		value -- essentially the expected polling duration -- against the actual polling duration
+		measured by comparing the "now" times reported in prepare and check.
+	 */
+	const struct timeval *timeout;
```
Comment (mergeconflict) [inline include/event2/event_watcher.h]: Note: this is the key bit of information we need to expose in this PR.

File: include/event2/event_watcher.h [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef EVENT2_EVENT_WATCHER_H_INCLUDED_
+#define EVENT2_EVENT_WATCHER_H_INCLUDED_
+
+/** @file event2/event_watcher.h
+
+	"Prepare" and "check" watchers. A "prepare" watcher is a callback that fires immediately before
```
Comment (mergeconflict) [inline include/event2/event_watcher.h]: If this PR is merged, how do we go about updating the user docs (e.g. http://www.wangafu.net/~nickm/libevent-book)?

**Comment by mergeconflict** [comment]
cc @htuch

File: evwatcher.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+event_watcher_check_new(struct event_base *base, check_watcher_cb callback)
+{
+	union event_watcher_cb cb = { .check = callback };
+	return event_watcher_new(base, cb, EVENT_WATCHER_CHECK_TYPE);
+}
+
+struct event_base *
+event_watcher_base(struct event_watcher *watcher)
+{
+	return watcher->base;
```
Comment (mergeconflict) [inline evwatcher.c]: Needs test coverage.

File: evwatcher.c [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+#include "event2/event_watcher.h"
+#include "event-internal.h"
+
+static inline struct event_watcher *
+event_watcher_new(struct event_base *base, union event_watcher_cb callback, unsigned type)
+{
+	struct event_watcher *watcher = mm_malloc(sizeof(struct event_watcher));
+	if (!watcher) {
+		return NULL;
```
Comment (mergeconflict) [inline evwatcher.c]: Coveralls notes this line lacks test coverage... Do we have any way of mocking mm_malloc to test the behavior when allocation fails?

File: CMakeLists.txt (hunk 1 above)
Comment (azat) [reply CMakeLists.txt]: autotools (which is primary build system for libevent right now) -- `Makefile.am`

*P.S. I do not ask you to fix manually written files for nmake, since I'm going to remove them*

File: evwatcher.c (hunk 2 above)
Comment (azat) [reply evwatcher.c]: yep

File: include/event2/event_watcher.h (hunk 5 above)
Comment (azat) [reply include/event2/event_watcher.h]: That will be great! Will you? (and also one good example will be great, like duration histograms that was mentioned in #710)

File: evwatcher.c (hunk 7 above)
Comment (azat) [reply evwatcher.c]: Sure, `event_set_mem_functions()`

File: include/event2/event_watcher.h (hunk 3 above)
Comment (azat) [reply include/event2/event_watcher.h]: I would prefer to drop this from this structure

Comment (azat) [reply include/event2/event_watcher.h]: BTW why it has been added here? This prepare/check thing does not looks like it is care about cached time, or maybe I'm missing something?

File: .gitignore [hunk 8]
```diff
@@ -5,6 +5,7 @@
 .#*
 *~
 *.swp
+.vscode
```
Comment (azat) [inline .gitignore]: I would prefer if this line will be in a separate patch (but keep it in this merge)
The main reason this is to keep main patch contains only prepare/check bits

File: event-internal.h [hunk 9]
```diff
@@ -205,6 +206,31 @@ struct event_once {
 	void *arg;
 };
 
+#define EVENT_WATCHER_PREPARE_TYPE 0
+#define EVENT_WATCHER_CHECK_TYPE   1
+#define EVENT_WATCHER_TYPES        2
```
Comment (azat) [inline event-internal.h]: Can we name it something like `EVENT_WATCHES_TYPE_MAX` or something similar, it makes meaning of this variable cleaner (that this is just the last of the watches defines)

File: event.c [hunk 10]
```diff
@@ -939,6 +946,15 @@ event_base_free_(struct event_base *base, int run_finalizers)
 	EVTHREAD_FREE_LOCK(base->th_base_lock, 0);
 	EVTHREAD_FREE_COND(base->current_event_cond);
 
+	/* Free all event watchers */
+	for (i = 0; i < EVENT_WATCHER_TYPES; ++i) {
+		while (!TAILQ_EMPTY(&base->event_watchers[i])) {
+			watcher = TAILQ_FIRST(&base->event_watchers[i]);
+			TAILQ_REMOVE(&base->event_watchers[i], watcher, next);
+			free(watcher);
```
Comment (azat) [inline event.c]: `mm_free` here too

File: event-internal.h (hunk 9 above)
Comment (azat) [reply event-internal.h]: also can you come up with more compact name, for example how about `EVW_PREPARE` (but this is just an advice and maybe it is too compact)

File: evwatcher.c [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (azat) [inline evwatcher.c]: I think that this header can be ommited... or just remove first two copyrights

File: include/event2/event_watcher.h [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+#endif
+#include <time.h>
+
+#include <event2/visibility.h>
+
+struct event_base;
+struct event_watcher;
+
+/**
+	Contextual information passed from event_base_loop to the "prepare" watcher callbacks. We define
```
Comment (azat) [inline include/event2/event_watcher.h]: Some of your lines are pretty long (not only here), can you make it the same style as include/event2/event.h  (although we do not have any strict restrictions on line length, unless it is *sane*)

File: include/event2/event_watcher.h [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+	this as a struct rather than individual parameters to the callback function for the sake of future
+	extensibility.
+ */
+struct prepare_watcher_cb_info {
```
Comment (azat) [inline include/event2/event_watcher.h]: - this does not contains `event_` prefix (and not only this one)
- but please do not call it `event_prepare_watcher_cb_info`, tooooo long
- and if this `struct` exposed then you will not have ability to change it, maybe it is better to hide it in module part and provide getter for the only field that it has, in this case user will not allow to change our timeout (since it has a pointer to our internal variable) which is not good from some point of vew

File: test/test-prepare-check.c [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+static int iteration = 0;
+static int prepare_callback_1_count = 0;
+static int prepare_callback_2_count = 0;
+static int check_callback_1_count = 0;
+static int check_callback_2_count = 0;
+static struct timeval start_time = { 0, 0 };
+static struct timeval end_time = { 0, 0 };
+
+#define log_check_callback \
```
Comment (azat) [inline test/test-prepare-check.c]: can you wrap this to `do { ... } while (0);` or at least `{}` block? (like in other places)

File: test/test-prepare-check.c [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+	start_time.tv_sec = 0;
+	event_watcher_free(prepare_callback_2_watcher);
+	event_watcher_free(check_callback_2_watcher);
+	event_base_once(base, -1, EV_TIMEOUT, &timeout_callback, 0, &tv);
+	event_base_dispatch(base);
+
+	/* this will free the watchers that weren't already freed in timeout_callback */
+	event_base_free(base);
+	return (0);
+}
```
Comment (azat) [inline test/test-prepare-check.c]: lines are also pretty long

File: test/test-prepare-check.c [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	iteration = 1;
+	start_time.tv_sec = 0;
+	event_watcher_free(prepare_callback_2_watcher);
+	event_watcher_free(check_callback_2_watcher);
+	event_base_once(base, -1, EV_TIMEOUT, &timeout_callback, 0, &tv);
+	event_base_dispatch(base);
+
+	/* this will free the watchers that weren't already freed in timeout_callback */
+	event_base_free(base);
+	return (0);
```
Comment (azat) [inline test/test-prepare-check.c]: please do not wrap the value (personally I think that this style came from some parts that had been migrated from other places, and that was a long time ago)

Comment (azat) [reply test/test-prepare-check.c]: so just `return 0;`

File: include/event2/event_watcher.h (hunk 5 above)
Comment (azat) [reply include/event2/event_watcher.h]: BTW the book is [here](https://github.com/nmathewson/libevent-book)

Comment (azat) [reply include/event2/event_watcher.h]: (but let's deal with the patches first)

File: include/event2/event_watcher.h (hunk 3 above)
Comment (mergeconflict) [reply include/event2/event_watcher.h]: It was mainly for convenience, not strictly necessary. I included it because one of the metrics we'd like to track is the duration between a `check` callback and the following `prepare` callback (on the next iteration of the event loop). But I know we can get that another way. The only really necessary bit is the `timeout` below, which there doesn't seem to be any other way to get.

File: include/event2/event_watcher.h (hunk 5 above)
Comment (mergeconflict) [reply include/event2/event_watcher.h]: Sounds good, thanks!

File: evwatcher.c (hunk 11 above)
Comment (mergeconflict) [reply evwatcher.c]: Ok, I'll take off the two copyrights from the three new files... Let me know if I misunderstood.

File: include/event2/event_watcher.h (hunk 12 above)
Comment (mergeconflict) [reply include/event2/event_watcher.h]: Sure, no problem. Out of curiosity, what tab-width do you use in your editor? I was using a tab-width of 2, but now I'm guessing that's much smaller than yours :)

Comment (azat) [reply include/event2/event_watcher.h]: Mine is four, but I guess that lots of code had been written with 6

**Comment by mergeconflict** [comment]
@azat I still haven't added the example under `sample/` you asked for -- I'll work on that next...

File: event.c [hunk 17]
```diff
@@ -1987,6 +2003,14 @@ event_base_loop(struct event_base *base, int flags)
 
 		event_queue_make_later_events_active(base);
 
+		/* Invoke prepare watchers before polling for events */
+		EVBASE_RELEASE_LOCK(base, th_base_lock);
```
Comment (mergeconflict) [inline event.c]: Another question: when my test was a standalone executable, it seemed there was no problem calling `event_base_gettimeofday_cached()` from within a watcher callback. But when I converted it into `regress` testcases, it broke: the lock was already held in the call, and it failed:

```
[err] /usr/local/google/home/mergeconflict/github/libevent/evthread.c:239: Assertion lock->count == 1 failed in evthread_debug_lock_mark_locked

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt    
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff66453fa in __GI_abort () at abort.c:89
#2  0x00007ffff69e49bf in event_exit (errcode=-559030611) at /usr/local/google/home/mergeconflict/github/libevent/log.c:96
#3  0x00007ffff69e4fb4 in event_errx (eval=-559030611, fmt=0x7ffff69e8774 "%s:%d: Assertion %s failed in %s")
    at /usr/local/google/home/mergeconflict/github/libevent/log.c:153
#4  0x00007ffff69dd5e4 in evthread_debug_lock_mark_locked (mode=0, lock=0x715560)
    at /usr/local/google/home/mergeconflict/github/libevent/evthread.c:239
#5  0x00007ffff69dcea4 in debug_lock_lock (mode=0, lock_=0x715560) at /usr/local/google/home/mergeconflict/github/libevent/evthread.c:261
#6  0x00007ffff69cd683 in event_base_gettimeofday_cached (base=0x7150e0, tv=0x7fffffffda30)
    at /usr/local/google/home/mergeconflict/github/libevent/event.c:445
```

Releasing the lock before invoking the watchers fixes the issue, and it seemed like the right thing to do since we also release it when invoking user callbacks for active events, but I want to double-check with you that this makes sense.

Comment (mergeconflict) [reply event.c]: I also wonder why it worked at all when my test was a standalone executable...

Comment (azat) [reply event.c]: > when my test was a standalone executable, it seemed there was no problem calling event_base_gettimeofday_cached() from within a watcher callback

libevent do not thread safe by default (since this is overhead), you can enable it with `evthread_use_pthreads()` this is the first part, see below for the second

> But when I converted it into regress testcases, it broke: the lock was already held in the call, and it failed:

Second part is that `regress` enables lock debugging (`evthread_enable_lock_debugging()`) and that is why this assert pops up (you can disable it by installing `EVENT_NO_DEBUG_LOCKS` environment variable) instead of dead locking the program (since relocking of the non-recursive mutex is most of the time this is deadlock, but strictly speaking it is "undefined behavior")

>Releasing the lock before invoking the watchers fixes the issue, and it seemed like the right thing to do since we also release it when invoking user callbacks for active events, but I want to double-check with you that this makes sense

Indeed, my mistake that I didn't noticed this during looking into this yesterday (actually I didn't look for the correctness that time, just "conceptually")
Any callback that is called from the event loop should not hold the base mutex.

File: include/event2/event_watcher.h [hunk 18]
```diff
@@ -29,98 +26,63 @@
...
+	Register a new "prepare" watcher, to be called in the event loop prior
+	to polling for events. Watchers will be called in the order they were
+	registered.
 	@param base the event_base to operate on.
 	@param callback the callback function to invoke.
 	@return a pointer to the newly allocated event watcher.
  */
 EVENT2_EXPORT_SYMBOL
-struct event_watcher *event_watcher_prepare_new(struct event_base *base, prepare_watcher_cb callback);
+struct event_watcher *event_watcher_prepare_new(struct event_base *base, event_prepare_cb callback);
```
Comment (azat) [inline include/event2/event_watcher.h]: looking into this again, and priorities support for this feature came to my mind, i.e. if we want this be generic, since if some library has event_base and registered watches and user want to call something before that library watches he cannot do this right now.

although this is pretty specific thing, and to avoid delaying merging this, we can at least reserve argument in the API for this (to make further changes ABI compatible) and return an error if `prio != 0`

File: test/regress_watcher.c [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+	/* check_callback_2 should only fire on the first iteration */
+	tt_int_op(iteration, ==, 0);
+end:
+	;
+}
+
+static void
+timeout_callback(evutil_socket_t fd, short events, void *arg)
+{
+	printf("timeout_callback\n");
```
[the remark this answers is not in the document]
Comment (mergeconflict) [reply test/regress_watcher.c]: I actually think it's best for me to just remove these - they don't really indicate success or failure, they were just helpful debugging info for me while I was learning how the event loop works.

File: event-internal.h [hunk 20]
```diff
@@ -206,16 +206,38 @@ struct event_once {
...
+	watcher callbacks. We define this as a struct rather than individual
+	parameters to the callback function for the sake of future
+	extensibility.
+ */
+struct event_prepare_cb_info {
+	/**
+		The timeout duration passed to the underlying implementation's `dispatch`.
+		See event_prepare_get_timeout.
+	 */
+	const struct timeval *timeout;
```
[the remark this answers is not in the document]
Comment (mergeconflict) [reply event-internal.h]: Hmmm. I don't feel strongly about this, but let me argue a little bit: I was thinking that the `const` provides enough safety -- if somebody casts away the `const`, it's their own fault. And maybe it's not actually a bad thing for users to modify the timeout? I don't know :)

Anyway, if you disagree, then I'm certainly ok with copying the timeout in the getter. The only trick is that the timeout may be `NULL` (e.g. if we're not blocking or we have no timeout events), so I'd need the getter to look something like:

```
int
event_prepare_get_timeout(const struct event_prepare_cb_info *info, struct timeval *timeout) {
	if (info->timeout) {
		*timeout = *(info->timeout);
		return 1;
	}
	return 0;
}
```

Does that look right?

File: include/event2/event_watcher.h (hunk 18 above)
Comment (mergeconflict) [reply include/event2/event_watcher.h]: Let me suggest an alternative approach: if we want to support priority for watchers later, we could add a function `event_watcher_priority_set`, similar to the existing `event_priority_set`. That would keep the `event_watcher_*_new` APIs minimal, and allow you to change the priority of a watcher...

File: test/regress_watcher.c [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+	  - Watchers should be called in the order they were registered.
+	  - Prepare watchers should be called before check watchers.
+	  - Freeing a watcher will stop callbacks to it, but not to other
+	    watchers.
+	  - Reported durations should align with the registered timeouts.
+	  - It should be possible to call back into libevent from a callback
+	    without a recursive lock.
+	  - If this test is compiled with ASAN or similar, this test also
+	    illustrates that event_base_free will free any watchers not
+	    previously freed by event_watcher_free.
```
[the remark this answers is not in the document]
Comment (mergeconflict) [reply test/regress_watcher.c]: I took out the extra two spaces but left the `-` bullets... let me know if that looks ok to you.

File: event.c (hunk 17 above)
Comment (mergeconflict) [reply event.c]: Thinking a bit about overhead... I'm wondering if we should gate this new feature behind a compilation flag, like:
```C
#ifdef EVENT__ENABLE_WATCHERS
...
#endif
```
so that you don't pay any performance cost (especially around locking/unlocking) if you're not going to use the feature. What do you think?

**Comment by azat** [comment]
> Thinking a bit about overhead... I'm wondering if we should gate this new feature behind a compilation flag, like:

The overhead should be neglible (one more lock/unlock, BTW we can get
rid off the impact by checking if there are any watchers and if not we
can skip lock/foreack/unlock, but again I don't think that even this
simple condition worth it) since event_base_loop() should not be called
tons of times (by those who cares about performance).

Also introducing yet another compilation option increases coverage
matrix (and it is already huge enough).

**Comment by azat** [comment]
> Hmmm. I don't feel strongly about this, but let me argue a little bit: I was thinking that the `const` provides enough safety -- if somebody casts away the `const`, it's their own fault. And maybe it's not actually a bad thing for users to modify the timeout? I don't know :)

I'm not sure that we should allow changing this time, and even if we
should, then explicitly via some API, like
event_prepare_set_{loop/base}_timeout() not via hackish casts
(personally I, don't care about architectures other then x86/amd64, but
this const-cast-away can break something for them).

But maybe I'm too suspicious about this...

> int
> event_prepare_get_timeout(const struct event_prepare_cb_info *info, struct timeval *timeout) {
> 	if (info->timeout) {
> 		*timeout = *(info->timeout);
> 		return 1;
> 	}
> 	return 0;
> }

Looks good.

File: include/event2/event_watcher.h (hunk 18 above)
Comment (azat) [reply include/event2/event_watcher.h]: Yep, sounds great

File: test/regress_watcher.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (azat) [reply test/regress_watcher.c]: I'm talking about that "common" comments style is one of the following:

- 4 spaces for alignment

```c
/** Brief description.

    foo bar

    @return 0 on success, -1 on failure
 */
```

- uses `*` at the beginning of the line
```c
/** Brief description
 *
 * Foo bar
 *
 * @return 0 on success, -1 on failure
 */
```

- but what I hadn't seen is using tabs for multi-line comments

File: event.c [hunk 22]
```diff
@@ -737,6 +735,11 @@ event_base_new_with_config(const struct event_config *cfg)
 		event_base_start_iocp_(base, cfg->n_cpus_hint);
 #endif
 
+	/* initialize event watcher lists */
+	for (i = 0; i < EV_WATCHER_MAX; ++i) {
```
Comment (azat) [inline event.c]: style: no `{}` for oneline body

i.e.

- correct
```c
for (;;)
    bar();
```

- not common:
```c
for (;;) {
    bar();
}
```

**Comment by mergeconflict** [comment]
I'm still working on the timing histogram example. Should I just add that in a separate pull request?

**Comment by mergeconflict** [comment]
Ok... I've added an example of how we might use prepare and check watchers to collect timing statistics. There's a fun implementation of constant-space histograms in there... it might be nice to turn that into a proper library at some point. @htuch, @mattklein123 - maybe you'd be interested in having a look at the example as well.

I haven't finished addressing all the style comments yet. I'll make another pass tomorrow.

**Comment by mergeconflict** [comment]
Hi @azat, thanks for all your feedback! I think I've addressed all the style issues, I've added a "what's new" entry, and I've updated the issue description.

I just want to double-check -- you want me to `git rebase -i` locally to squash all the commits, using this issue description as the commit message, and then force-push the branch -- is that right?

**Comment by azat** [comment]
> I think I've addressed all the style issues, I've added a "what's new" entry, and I've updated the issue description.

Thanks! Will take a closer look today or at least tomorrow

> I just want to double-check -- you want me to git rebase -i locally to squash all the commits, using this issue description as the commit message, and then force-push the branch -- is that right?

Yep git rebase locally, squash all the commits, update commit
description and git push --force

File: evwatcher.c [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+static inline struct event_watcher *
+event_watcher_new(struct event_base *base, union event_watcher_cb callback, unsigned type)
+{
+	struct event_watcher *watcher = mm_malloc(sizeof(struct event_watcher));
+	if (!watcher)
+		return NULL;
+	watcher->base = base;
+	watcher->type = type;
+	watcher->callback = callback;
+	TAILQ_INSERT_TAIL(&base->watchers[type], watcher, next);
```
Comment (azat) [inline evwatcher.c]: missing base lock/unlock

File: evwatcher.c [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+struct event_base *
+event_watcher_base(struct event_watcher *watcher)
+{
+	return watcher->base;
+}
+
+void
+event_watcher_free(struct event_watcher *watcher)
+{
+	TAILQ_REMOVE(&watcher->base->watchers[watcher->type], watcher, next);
```
Comment (azat) [inline evwatcher.c]: missing base lock/unlock

File: sample/watcher-timing.c [hunk 25]
```diff
@@ -0,0 +1,315 @@
...
+  [2] https://github.com/mergeconflict/histogram
+ */
+
+/** Compare two doubles for equality without the compiler warning. This is
+ * probably the wrong thing to do, but this is just sample code :) */
+static inline int
+eq(double a, double b)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wfloat-equal"
```
Comment (azat) [inline sample/watcher-timing.c]: can you show that this is required?

File: sample/watcher-timing.c [hunk 26]
```diff
@@ -0,0 +1,315 @@
...
+		 h->bins[h->gap + 1].centroid * h->bins[h->gap + 1].count) /
+		(h->bins[h->gap].count + h->bins[h->gap + 1].count);
+	h->bins[h->gap + 1].count += h->bins[h->gap].count;
+}
+
+static double
+histogram_query(const struct histogram *h, double quantile)
+{
+	unsigned lhs = 0, rhs = 0;
+	struct bin lhs_bin, rhs_bin;
```
Comment (azat) [inline sample/watcher-timing.c]: can you initialize them, to avoid `[-Wmaybe-uninitialized]` ? (you can see this warnings with release build type)

Comment (azat) [reply sample/watcher-timing.c]: hm, and looks like this is not false-positive

File: evwatcher.c [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+	mm_free(watcher);
+}
+
+int
+event_prepare_get_timeout(const struct event_prepare_cb_info *info, struct timeval *timeout)
```
Comment (azat) [inline evwatcher.c]: hm, looking into this again, and `event_prepare_get_timeout`does not looks like a good naming, how about `evwatch_prepare_get_timeout()` ? (and convert the whole patch into this)

File: sample/watcher-timing.c [hunk 28]
```diff
@@ -0,0 +1,315 @@
...
+	struct timeval one_second = { 1, 0 };
+
+	delays = histogram_new(100);
+	durations = histogram_new(100);
+
+	event_add(ev, &one_second);
+	event_watcher_prepare_new(base, &on_prepare);
+	event_watcher_check_new(base, &on_check);
+	event_base_dispatch(base);
+
```
Comment (azat) [inline sample/watcher-timing.c]: all below (`*free()`) will never be reached, but what you can do is add a handler for `SIGINT` (see `evsignal_*()`) and call `event_base_loopexit()` on it, then you will got no leaks on termination by interrupting.

File: sample/watcher-timing.c (hunk 25 above)
Comment (mergeconflict) [reply sample/watcher-timing.c]: I get the following warning without it:
```
sample/watcher-timing.c:43:11: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
        return a == b;
               ~ ^  ~
```

**Comment by azat** [comment]
> +#pragma GCC diagnostic ignored "-Wfloat-equal"
>
> I get the following warning without it:
>
> sample/watcher-timing.c:43:11: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
>         return a == b;
>                ~ ^  ~

Got it, thanks! (yet another difference between autotools and cmake,
will fix this later).

File: evwatcher.c (hunk 27 above)
Comment (mergeconflict) [reply evwatcher.c]: Do you mean to change all:
```C
struct event_watcher => struct evwatch
struct event_prepare_cb_info => struct evwatch_prepare_cb_info
struct event_check_cb_info => struct evwatch_check_cb_info

event_prepare_cb => evwatch_prepare_cb
event_check_cb => evwatch_check_cb

event_watcher_prepare_new => evwatch_prepare_new
event_watcher_check_new => evwatch_check_new
event_watcher_base => evwatch_base
event_watcher_free => evwatch_free
```
Should I also rename `event_watcher.h` to `evwatch.h`? I don't understand the naming conventions here... Or do you just want me to rename the one function?

Comment (azat) [reply evwatcher.c]: Yep, all of them, i.e. like `evbuffer` was done
And if we are doing this like `evbuffer`, then the file should be called `<event2/watch.h>`

**Comment by mergeconflict** [comment]
@azat updated and rebased. The lock/unlock was a good catch. Thanks again for the thorough review!

**Comment by mergeconflict** [comment]
Hi @azat, everything's updated. Let me know if there's anything else!

**Comment by azat** [comment]
@mergeconflict everything looks good!

And while thinking about this one more time, I think that user-passed argument will not be superfluous, i.e. `evwatch_prepare_new(struct event_base *base, evwatch_prepare_cb callback, void *arg)`, can you add this? (and I think that this is the last bit, sorry for not noticing this before)

**Comment by mergeconflict** [comment]
> And while thinking about this one more time, I think that user-passed argument will not be superfluous, i.e. `evwatch_prepare_new(struct event_base *base, evwatch_prepare_cb callback, void *arg)`, can you add this? (and I think that this is the last bit, sorry for not noticing this before)

Sure, no problem. I was also actually thinking it might be handy. I'll put the `*arg` as an additional argument in the callbacks, after the `evwatch_*_cb_info` to stay consistent with how the `evbuffer` callbacks are designed.

**Comment by azat** [comment]
So everything looks great, will apply this once the CI will be passed (or almost passed)
Thanks for the quick reaction!

**Comment by azat** [comment]
One more thing pops up `evwatch_check_cb_info` is empty, and while this is accepted by gcc it is not strictly valid and VS do not pass this:
```
c:\projects\libevent\event-internal.h(221): error C2016: C requires that a struct or union has at least one member [C:\projects\libevent\build-cmake\event_core_shared.vcxproj]
```

Maybe this can be suppressed with some kind of pragma, but I think that better option will be to add one field there instead, like the time the dispatch took

**Comment by azat** [comment]
And commit message should be formatted appropriate, i.e.:

```
subject
<   blank line   >
body
```

since otherwise git will think that the whole commit message is the subject (without empty line)

(first link from google https://chris.beams.io/posts/git-commit/#seven-rules)

**Comment by mergeconflict** [comment]
> Maybe this can be suppressed with some kind of pragma, but I think that better option will be to add one field there instead, like the time the dispatch took

Interesting, that's actually very helpful for us -- it would mean that we don't have to save the prepare and check timestamps externally. I'm thinking to rename the prepare timeout `expected_timeout` and call the check timeout `actual_timeout`, so that it's obvious to users that you can compare the actual and expected values. What do you think?

**Comment by azat** [comment]
>Interesting, that's actually very helpful for us -- it would mean that we don't have to save the prepare and check timestamps externally. I'm thinking to rename the prepare timeout expected_timeout and call the check timeout actual_timeout, so that it's obvious to users that you can compare the actual and expected values. What do you think?

I like the idea in general, but don't like such long fields. I guess it is fine here

**Comment by mergeconflict** [comment]
Ok... I've implemented timing, but I haven't squashed the commits yet, since I'm not sure if you'll be ok with it. Specifically, to get accurate timing, I've had to add two `evutil_gettimeofday` calls before and after `evsel->dispatch`, which might be unacceptable overhead in some cases. I tried using the cached time, but it was too inaccurate (like 3000us compared to about 1000us expected).

Of course I'd be very happy to keep this implementation, since it supports my use case directly, but I understand if it's not good for others. If that's the case, then we'll solve the Win32 empty struct issue by simply adding a dummy unused field instead.

**Comment by mergeconflict** [comment]
Hmm... I just noticed I have a link error on Windows:
```
LINK : fatal error LNK1104: cannot open file 'm.lib' [C:\projects\libevent\build-cmake\watch-timing.vcxproj]
```
Trying a fix...

**Comment by azat** [comment]
>to get accurate timing, I've had to add two evutil_gettimeofday calls before and after evsel->dispatch, which might be unacceptable overhead in some cases

Well if we are talking about modern linux (but maybe I missed something, since this is 5min experience) then I don't think that cached version is worth it, and just out of curiosity I wrote [simple test](https://gist.github.com/azat/2168ee8cc3a29397be63dcf9894f7075#file-le-gettimeofday-c) and this assumption is correct (actually non cached version is a little bit faster)

But we have lots of other supported architectures, so I let's use cached version here and this will be addressed later

(because since we do have this `event_base_gettimeofday_cached()` now we should use it, if we do not want it, it should be removed completely in one pass)

**Comment by mergeconflict** [comment]
> But we have lots of other supported architectures, so I let's use cached version here and this will be addressed later

The issue with using the cached time is that it's too inaccurate, at least for the main use case we care about in Envoy. The goal of measuring the actual vs. expected polling duration is that if you compare the two, the actual duration should be *very close* to the expected duration - within one kernel timeslice usually. If the delay is larger, it's an indication that your kernel is swamped. Using the cached time will give the misleading impression that something's wrong...

My suggestion is, let's leave this out for now. In the future, if we get rid of the cached time (since your experimental results seem good), then we can put the accurate timing here. Does that sound ok?

**Comment by azat** [comment]
> My suggestion is, let's leave this out for now. In the future, if we get rid of the cached time (since your experimental results seem good), then we can put the accurate timing here. Does that sound ok?

Agree

**Comment by mergeconflict** [comment]
Ok, cool. I've updated:

* Revert the timing thing
* Fix the empty struct issue by adding a dummy unused field
* Fix the git commit message
* Squashed and rebased against the latest changes in master

**Comment by azat** [comment]
Applied, thanks for the effort!

**Comment by mergeconflict** [comment]
Thanks @azat!
